linera-execution/src/wasm/runtime_api.rs

```diff
@@ -791,34 +378,299 @@ where
+        caller
+            .user_data_mut()
+            .runtime
+            .try_call_application(authenticated, callee_id, argument)
+            .map_err(|error| RuntimeError::Custom(error.into()))
+    }
+
+    /// Adds an item to an event stream.
+    fn emit(
+        caller: &mut Caller,
+        name: StreamName,
+        key: Vec<u8>,
+        value: Vec<u8>,
+    ) -> Result<(), RuntimeError> {
+        caller
+            .user_data_mut()
+            .runtime
+            .emit(name, key, value)
+            .map_err(|error| RuntimeError::Custom(error.into()))
+    }
+
+    /// Queries a service and returns the response.
+    fn query_service(
+        caller: &mut Caller,
+        application_id: ApplicationId,
+        query: Vec<u8>,
+    ) -> Result<Vec<u8>, RuntimeError> {
+        caller
+            .user_data_mut()
+            .runtime
+            .query_service(application_id, query)
+            .map_err(|error| RuntimeError::Custom(error.into()))
+    }
+
+    /// Consume some fuel.
+    ///
+    /// This is intended for the metering instrumentation, but if the user wants to donate
+    /// some extra fuel, more power to them!
+    fn consume_fuel(caller: &mut Caller, fuel: u64) -> Result<(), RuntimeError> {
+        caller
+            .user_data_mut()
+            .runtime_mut()
+            .consume_fuel(fuel)
+            .map_err(|e| RuntimeError::Custom(e.into()))
+    }
+
+    /// Returns the round in which this block was validated.
+    fn validation_round(caller: &mut Caller) -> Result<Option<u32>, RuntimeError> {
+        caller
+            .user_data_mut()
+            .runtime_mut()
+            .validation_round()
+            .map_err(|error| RuntimeError::Custom(error.into()))
+    }
 
     /// Writes a batch of `operations` to storage.
     fn write_batch(
         caller: &mut Caller,
         operations: Vec<WriteOperation>,
     ) -> Result<(), RuntimeError> {
-        WriteBatch::write_batch(&mut caller.user_data_mut().runtime, Batch { operations })
+        caller
+            .user_data_mut()
+            .runtime_mut()
+            .write_batch(Batch { operations })
             .map_err(|error| RuntimeError::Custom(error.into()))
     }
 }
 
-// TODO(#1977): Remove once the WIT interface does not include `write-batch` in the service system
-// API
-/// An extension trait to separate the behavior between the contract runtime and the service
-/// runtime.
-pub trait WriteBatch {
-    /// Writes a [`Batch`] of operations to storage.
-    fn write_batch(&mut self, batch: Batch) -> Result<(), ExecutionError>;
-}
+/// An implementation of the system API made available to services.
+#[derive(Default)]
+pub struct ServiceRuntimeApi<Caller>(PhantomData<Caller>);
 
-impl WriteBatch for ContractSyncRuntimeHandle {
-    fn write_batch(&mut self, batch: Batch) -> Result<(), ExecutionError> {
-        ContractRuntime::write_batch(self, batch)
+#[linera_witty::wit_export(package = "linera:app")]
```

Comment:
Nit: Same here regarding the path prefix.

```suggestion
#[wit_export(package = "linera:app")]
```